bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+ */
+std::string RowRange::GetEndKey() const {
+  std::string result = "";
+
+  switch (range_.end_key_case()) {
+    case btproto::RowRange::kEndKeyOpen:
+      result = range_.end_key_open();
+      break;
+    case btproto::RowRange::kEndKeyClosed:
+      result = range_.end_key_closed();
+      break;
+    // Same as btproto::RowRange::START_KEY_NOT_SET.
+    case btproto::RowRange::END_KEY_NOT_SET:
+      break;
+  }
+  return result;
+}
+
+/**
+ * Finalize the end points for intersection.
+ * position : gives the point position (start or end).
+ */
+void RowRange::FixKeyPointAndTypeIntersect(RowRange::PositionType position,
+                                           std::string& result_key,
+                                           RowRange::KeyType& result_type,
+                                           std::string rhs_key,
+                                           RowRange::KeyType rhs_type) {
+  // Pick the max value for start point, and min value for the end.
+  bool comparison = (position == PositionType::START ? result_key < rhs_key
+                                                     : result_key > rhs_key);
+  if (comparison) {
+    result_key = rhs_key;
+    result_type = rhs_type;
+  } else if (result_key == rhs_key && KeyType::OPEN == rhs_type) {
+    // prefer OPEN type in case of a tie.
+    result_type = rhs_type;
+  }
+}
+
+/**
+ * Initialize the protocol Buffer within the RowRange.
+ */
+void RowRange::SetProtoBufRangePoint(RowRange::PositionType position,
+                                     RowRange& row_range, std::string point_key,
+                                     RowRange::KeyType point_type) {
+  if (position == PositionType::START) {
+    if (point_type == KeyType::OPEN)
+      row_range.range_.set_start_key_open(point_key);
+    else if (point_type == KeyType::CLOSED)
+      row_range.range_.set_start_key_closed(point_key);
+  } else {  // End point.
+    if (point_type == KeyType::OPEN)
+      row_range.range_.set_end_key_open(point_key);
+    else if (point_type == KeyType::CLOSED)
+      row_range.range_.set_end_key_closed(point_key);
+  }
+}
+
+std::string RowRange::debug_string() const {
+  KeyType start_type{GetStartKeyType()}, end_type{GetEndKeyType()};
+
+  std::string result = "";
+  switch (start_type) {
+    case KeyType::OPEN:
+      result = absl::StrCat("(", range_.start_key_open());
+      break;
+    case KeyType::CLOSED:
+      result = absl::StrCat("[", range_.start_key_closed());
+      break;
+    case KeyType::NOT_SET:
+      // For an unset start key, the value is "", inclusive.
+      result = "[";
+      break;
+  }
+  result = absl::StrCat(result, "..");
+  switch (end_type) {
+    case KeyType::OPEN:
+      result = absl::StrCat(result, range_.end_key_open(), ")");
+      break;
+    case KeyType::CLOSED:
+      result = absl::StrCat(result, range_.end_key_closed(), "]");
+      break;
+    case KeyType::NOT_SET:
+      result = absl::StrCat(result + "NOT_SET)");
```

Comment:
I think you wanted `result, "NOT_SET"` instead of `result + "NOT SET"`.

Reply:
Good catch.